R/dash.R

```diff
@@ -831,9 +832,46 @@ Dash <- R6::R6Class(
       if (is.null(private$callback_context_)) {
         warning("callback_context is undefined; callback_context may only be accessed within a callback.")
       }
+
       private$callback_context_
     },
 
+    # ------------------------------------------------------------------------
+    # request and return callback timing data
+    # ------------------------------------------------------------------------
+    #' @description
+    #' Records timing information for a server resource.
+    #' @details
+    #' The `callback_context.record_timing` method permits retrieving the
+    #' duration required to execute a given callback. It may only be called
+    #' from within a callback; a warning will be thrown and the method will
+    #' otherwise return `NULL` if invoked outside of a callback.
+    #' 
+    #' @param name Character. The name of the resource.
+    #' @param duration Numeric. The time in seconds to report. Internally, this is
+    #' rounded to the nearest millisecond.
+    #' @param description Character. A description of the resource.
+    #'
+    callback_context.record_timing = function(name,
+                                              duration=NULL, 
+                                              description=NULL) {
+      if (is.null(private$callback_context_)) {
+        warning("callback_context is undefined; callback_context.record_timing may only be accessed within a callback.")
+        return(NULL)
+      }
+
+      timing_information <- self$server$get_data("timing-information")
+
+      if (name %in% timing_information) {
+        stop(paste0("Duplicate resource name ", name, " found."), call.=FALSE)
+      }
+
+      timing_information[[name]] <- list("dur" = round(duration * 1000), 
+                                         "desc" = description)
+
+      self$server$set_data("timing-information", timing_information)
```

Comment:
This may be a dumb question from an R noob, but is this `set_data` necessary? Above you `get_data` into `timing_information`, then you add a new key to `timing_information`, but if it's still the same object, won't that new key already be reflected in the server data?

Reply:
@alexcjohnson Assuming I understand your question, yes, it is necessary. R uses by pass-by-value semantics for the most part, so even if we update `timing_information`, that object still lives in the scope of `callback_context.record_timing`.

If we don't use `set_data` to store the updated object, the new information is lost when the function exits.

This might be interesting (though I haven't read it carefully myself):

https://homepage.divms.uiowa.edu/~luke/R/references.html